### PR TITLE
Pagination fixes

### PR DESCRIPTION
**Pull request by ryantheleach**
This is the first commits of what will be several bug fixes to Pagination.

- [x] https://github.com/SpongePowered/SpongeCommon/issues/748
- [x] https://github.com/SpongePowered/SpongeCommon/issues/593
- [x] https://github.com/SpongePowered/SpongeCommon/issues/592
and possibly 
- [x] https://github.com/SpongePowered/SpongeCommon/issues/708 *
if the source can be determined.
- [x] https://github.com/SpongePowered/SpongeCommon/issues/744

* Should be fixed when this is merged into 1.9, partially a Mojang bug.

### CODE REVIEW

**Comment by kmecpp** [comment]
There's also #744

**Comment by ryantheleach** [comment]
Thanks

**Comment by ryantheleach** [comment]
I'm aware of the imports and unused class, will fix this correctly next time I work on this, wanted to commit for now in order to get comments. I think making the custom childCommandExecutor is cleaner then the unused class. https://github.com/SpongePowered/SpongeAPI/issues/1272

**Comment by ryantheleach** [comment]
getCharWidth now works 100% but there is still a bug with either the way the text builders construct the center text, or with the way the length of texts is calculated.

I will check back on this tomorrow, but would appreciate if someone more familar with text could take a look, as I am confident that the getWidth(int codepoint, boolean bold) is 100% correct.

File: src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java [hunk 1]
```diff
@@ -85,37 +89,89 @@ public int getLinesPerPage(MessageReceiver source) {
...
+     *
+     * @param text
+     * @return
+     */
+    public int getLines(Text text) {
+        return (int) Math.ceil((double) getLength(text) / LINE_WIDTH);
     }
 
-    private double getWidth(int codePoint, boolean isBold) {
+    private static class thing{
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java]: ..."thing"?

File: src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java [hunk 2]
```diff
@@ -85,37 +89,89 @@ public int getLinesPerPage(MessageReceiver source) {
...
+        private final char i;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java]: What do these 1-character fields represent? Giving them real names would be nice.

File: src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java (hunk 1 above)
Comment (ryantheleach) [inline src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java]: It's WIP changes that were pushed in a hurry, no where near ready

**Comment by ryantheleach** [comment]
https://github.com/SpongePowered/SpongeCommon/pull/790/commits/6be17eac9ac248076bbc57c49b25ad62f597d050

Was the last clean commit before things get muddy. If you want to review changes I'd recommend reviewing those or older, anything newer is/was subject to change or in flux.

File: src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java [hunk 3]
```diff
@@ -34,11 +36,13 @@
...
-import org.spongepowered.api.text.format.TextStyles;
+import org.spongepowered.common.interfaces.text.IMixinText;
+import org.spongepowered.common.text.ChatComponentIterable;
+import org.spongepowered.common.text.ChatComponentIterator;
+import org.spongepowered.common.text.SpongeTexts;
 
 import java.io.IOException;
-import java.util.Collections;
-import java.util.List;
+import java.util.*;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java]: No star imports

File: src/main/java/org/spongepowered/common/service/pagination/SpongePaginationService.java [hunk 4]
```diff
@@ -24,12 +24,18 @@
...
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.spongepowered.api.command.args.GenericArguments.firstParsing;
 import static org.spongepowered.api.command.args.GenericArguments.integer;
+import static org.spongepowered.api.command.args.GenericArguments.optional;
 import static org.spongepowered.common.util.SpongeCommonTranslationHelper.t;
 
-import com.google.common.collect.ImmutableList;
-import com.google.common.collect.Iterables;
-import com.google.common.collect.MapMaker;
+import com.google.common.collect.*;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/service/pagination/SpongePaginationService.java]: No star imports

File: src/main/java/org/spongepowered/common/service/pagination/SpongePaginationService.java [hunk 5]
```diff
@@ -130,6 +115,138 @@ SourcePaginations getPaginationState(MessageReceiver source, boolean create) {
...
+                } catch (final Throwable e) {
+                    throw new RuntimeException(e);
+                }
+            }
+
+            void acceptThrows(T t) throws CommandException;
+        }
+
+        private enum Page {
+            NEXT("next","n"), PREV("prev","p","previous"), PAGE("page");
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/service/pagination/SpongePaginationService.java]: Have each enum constant on its own line

File: src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java [hunk 6]
```diff
@@ -85,88 +89,169 @@ public int getLinesPerPage(MessageReceiver source) {
...
                 continue;
             }
-            boolean isBold = child.getStyle().contains(TextStyles.BOLD);
-            for (int i = 0; i < txt.length(); ++i) {
-                columnCount += getWidth(txt.codePointAt(i), isBold);
+
+            IChatComponent childComponents = SpongeTexts.toComponent(child);
+            Boolean bold = childComponents
+                    .getChatStyle()
+                    .bold;
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java]: You need to use `.getBold()` or however that is called to get the inherited style

**Comment by ryantheleach** [comment]
the legacyChatWIdth works better, but I still think it's got issues.

1. Unknown minor issues.
2. Translation issues, only works with translations available from the server side.
3. selectors / score text elements

I believe that it's working closer to what it should be however. I will try to get this cleaned up with the limited internet access over the next 2 weeks.

File: src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java [hunk 7]
```diff
@@ -85,88 +89,206 @@ public int getLinesPerPage(MessageReceiver source) {
...
-            final String txt;
+    /**
+     * compute the length of a given text
+     * @param text
+     * @return the number of character pixels/columns the line takes up
+     */
+    @VisibleForTesting
+    int getLength(Text text) {
+        //TODO: work out what the difference is between the legacy width, and getLength, and implement it in getLength.
+        if(true)
```
Comment (ryantheleach) [inline src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java]: I would prefer to not just copy the Mojang code blindly, and actually understand why getLegacyWidth works more accurately then what has been effectively short circuited here. 

If you are happy with the results however I guess It's fine to just prune the pagination out.

I'm also not entirely happy with the fact that PaginationCalculator now relies upon the text mixins as it makes it harder to unit test.

**Comment by ryantheleach** [comment]
Somewhere in the last few cleaning up commits I managed to regress that the equals in the center padding split onto a new line again. not ready to merge, even with legacy width.

File: src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java [hunk 8]
```diff
@@ -85,88 +91,206 @@ public int getLinesPerPage(MessageReceiver source) {
...
+            return getLegacyWidth(TextSerializers.LEGACY_FORMATTING_CODE.serialize(text));
+
+        IChatComponent component = SpongeTexts.toComponent(text);
+
+        Iterator<IChatComponent> it = component.iterator();
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java]: Use `IMixinChatComponent.withChildren()` and for loop

File: src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java [hunk 9]
```diff
@@ -85,88 +91,206 @@ public int getLinesPerPage(MessageReceiver source) {
...
+        int total = 0;
+
+        while(it.hasNext()){
+            PrimitiveIterator.OfInt i_it;
+            IChatComponent child = it.next();
+            if(child instanceof ChatComponentText) {
+                i_it = ((ChatComponentText) child).getUnformattedTextForChat().codePoints().iterator();
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java]: You can make this an OR and cast to `IChatComponent`

Comment (ryantheleach) [inline src/main/java/org/spongepowered/common/service/pagination/PaginationCalculator.java]: I originally thought that I was going to need to call 2 different methods, Thanks.

**Comment by ryantheleach** [comment]
Talking to minecrell in IRC helped discover the remaining issue, should be good to merge, if the code review goes well.

**Comment by Aaron1011** [comment]
@ryantheleach: This looks good to merge. Can you squash the commits into one?

**Comment by ryantheleach** [comment]
Not until this coming weekend. Still semi-without internet or IDE, or machines with git.

**Pull request merged**